File: validparan.py
class Solution:
    # TC-O(n),SC-O(n)
    def isValid(self, s: str) -> bool:
        dict = {'}':'{',')':'(',']':'['}
        stack=[]
        for i in s:
            if i in dict:
                if len(stack)>0 and stack[-1]==dict[i]:
                    stack.pop()
                else:
                    stack.append(i)
            else:
                stack.append(i)
            print(stack)
        if len(stack)==0:
            return True
        else:
            return False

        

File: ValidParan.java
// TC - O(N)
// SC - O(N)
//Here we are using stack. We will check if its opening bracket if it is, 
//then we will push corresponding closing bracket. 
class Solution {
    public boolean isValid(String s) {
       Stack<Character> st = new Stack<>();
       if((s.length()==0) || s == null) {
           return true;
       }

        for(int i=0; i<s.length(); i++){
            char c = s.charAt(i);
            if(c == '('){
                st.push(')');
            }else if(c == '['){
                st.push(']');
            }else if(c == '{'){
                st.push('}');
            }else if(st.isEmpty() || c != st.pop()){
                return false;
            }
        }
        return st.isEmpty();
    }   
}